src/sets/signerSet.ts

```diff
@@ -99,381 +96,224 @@ class SignerSet {
-
-    // If child is a custody address
-    if (this._custodyAddresses.has(childKey)) return err('SignerSet.add: child cannot be a custody signer');
-
-    // If (parent, child) exists in eAdds
-    const existingEdgeAddsHash = this._edgeAdds.get(edgeKey);
-    if (existingEdgeAddsHash) {
-      // Update eAdds with higher order hash
-      if (hashCompare(existingEdgeAddsHash, hash) < 0) {
-        this._edgeAdds.set(edgeKey, hash);
-      }
-      return ok(undefined);
-    }
-
-    // If (parent, child) exists in eRems
-    const existingEdgeRemovesHash = this._edgeRemoves.get(edgeKey);
-    if (existingEdgeRemovesHash) {
-      // Update eRems with higher order hash
-      if (hashCompare(existingEdgeRemovesHash, hash) < 0) {
-        this._edgeRemoves.set(edgeKey, hash);
-      }
-      return ok(undefined);
-    }
-
-    // If parent exists in vAdds or custodyAdds
-    if (this._vertexAdds.has(parentKey) || this._custodyAdds.has(parentKey)) {
-      // If child does not exist in vertices
-      if (!this._vertices.has(childKey)) {
-        // Add child to vAdds and (a,b) to eAdds
-        this._vertexAdds.add(childKey);
-        this._edgeAdds.set(edgeKey, hash);
-
-        return ok(undefined);
-      }
-
-      // If child exists in vAdds
-      else if (this._vertexAdds.has(childKey)) {
-        // For each edge (*, child) in edgeAdds (though there should be only one parent)
-        const parentEdges = this._edgeAddsByChild.get(childKey) || new Set();
-        for (const existingEdgeKey of parentEdges) {
-          const existingEdgeHash = this._edgeAdds.get(existingEdgeKey);
-          if (!existingEdgeHash) return err('SignerSet.add: unexpected state');
-
-          // If existing message wins
-          if (hashCompare(existingEdgeHash, hash) > 0) {
-            // Add (parent, child) to eRems
-            this._edgeRemoves.set(edgeKey, hash);
+  private mergeSignerAdd(message: SignerAdd): Result<void, string> {
+    const custodyAddress = this.sanitizeKey(message.signer);
+    const signerKey = this.sanitizeKey(message.data.body.childKey);
+
+    // If custody address has been removed, no-op
+    if (this._custodyRemoves.has(custodyAddress)) return ok(undefined);
+
+    // If custody address is missing, fail
+    const custodyAddEvent = this._custodyAdds.get(custodyAddress);
+    if (!custodyAddEvent) return err('SignerSet.mergeSignerAdd: custodyAddress does not exist');
+
+    // If signer add exists
+    const existingSignerAdd = this._signerAdds.get(signerKey);
+    if (existingSignerAdd) {
+      const existingCustodyAddEvent = this._custodyAdds.get(this.sanitizeKey(existingSignerAdd.signer));
+
+      if (existingCustodyAddEvent) {
+        // If existing block number wins, no-op
+        if (existingCustodyAddEvent.blockNumber > custodyAddEvent.blockNumber) return ok(undefined);
+
+        // If block numbers are the same
+        if (existingCustodyAddEvent.blockNumber === custodyAddEvent.blockNumber) {
```

Comment:
is it possible to create a contract that could shuffle around an id between two addresses multiple times in the same transaction? and would that cause a problem here?

Reply:
this logic is particularly hard to parse as a reader -- i think its just very tricky to understand and might need lots of docs

Reply:
Adding comments and test to confirm that a CustodyRemoveAll message will only remove custody addresses added in a prior block. Custody addresses added in the same block (regardless of the order Ethereum decides for them), will be preserved until a custody address removes them from a later block.